awswrangler/_utils.py

```diff
@@ -65,6 +67,58 @@
 
 Boto3PrimitivesType = Dict[str, Optional[str]]
 
+# A mapping from import name to package name (on PyPI) for packages where
+# these two names are different.
+INSTALL_MAPPING = {
+    "redshift_connector": "redshift",
+    "pymysql": "mysql",
+    "pg8000": "postgres",
+    "pyodbc": "sqlserver",
+    "gremlin_python": "gremlin",
+    "opensearchpy": "opensearch",
+}
+
+
+def check_optional_dependency(
+    module: Optional[ModuleType],
+    name: str,
+) -> Callable[..., Any]:
```

Comment:
I wonder if we could use a generic here, similar to [this](https://github.com/aws/aws-sdk-pandas/blob/0e7710e67f0cf21917c4a85170112241e7306ae0/awswrangler/distributed/ray/_pool.py#L19). This way, the function signature won't be lost because of the annotation for the purposes of type checking.

Reply:
A bit tricky given that the function is wrapped inside not only a decorator but also a decorator factory. Moreover the return type in your example is very specific List[MapOutput] while this decorator factory wraps multiple functions with different return types.

Do you have suggestions on the generic to use?

Reply:
I was able to solve this with the following changes:

```python
def check_optional_dependency(
    module: Optional[ModuleType],
    name: str,
) -> Callable[[FunctionType], FunctionType]:
    def decorator(func: FunctionType) -> FunctionType:
        @wraps(func)
        def inner(*args: Any, **kwargs: Any) -> Any:
            if not module:
                package_name = INSTALL_MAPPING.get(name)
                install_name = package_name if package_name is not None else name
                raise ModuleNotFoundError(
                    f"Missing optional dependency '{name}'. " f"Use pip awswrangler[{install_name}] to install it."
                )
            return func(*args, **kwargs)

        return cast(FunctionType, inner)

    return decorator
```